digest/src/newtype.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{
+    core_api::BlockSizeUser,
+    crypto_common::hazmat::{DeserializeStateError, SerializableState, SerializedState},
+    CustomizedInit, FixedOutput, FixedOutputReset, HashMarker, Output, OutputSizeUser, Reset,
+    Update,
+};
+use core::{fmt, marker::PhantomData};
+
+// https://github.com/taiki-e/pin-project/issues/102#issuecomment-540472282
+#[doc(hidden)]
+#[derive(Clone, Default)]
+pub struct Wrapper<'a, T> {
```

Comment:
Sorry, but I still don't quite understand the relevance of this wrapper. Why can't the macro implement the struct as a direct thin wrapper around `CoreCtWrapper` and other wrapper types?

Reply:
Why do you need to deal with lifetimes here?

This snippet works just fine: https://play.rust-lang.org/?gist=9f86f1ec8c12cbb176cb8e1cce852e56

Reply:
Yes, but this doesn't work anymore: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=0c16e92508e7903da2d6099d956f57dd

This is what the lifetime gives me: a bound on the trait implementation of the wrapper.

Reply:
I do not understand. Why do you expect for `wrapper!(String => Email);` to work in you snippet?

Reply:
Because the inner type may or may not implement Foo, and I want the wrapper to implement Foo only if the inner implements it.

This is what the lifetime hack gives me.

Reply:
Ah, I see. It's certainly a neat trick, but it's a bit too much magic for my taste. The macro itself already  obscures the implementation a bit, so I would prefer to have something more straightforward.

We can either have macros for common scenarious (e.g. `impl_fixed_digest!`) or list macros which we want to implement using the wrapped type explicitly. I am not sure which solution would be better, so I will need to experiment and think about it a bit.

Reply:
alright well I'll pause a bit then :)